core/src/main/scala/fs2/job/JobManager.scala

```diff
@@ -0,0 +1,214 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fs2
+package job
+
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+
+import cats.effect.{Concurrent, Timer}
+
+import fs2.concurrent.{Queue, SignallingRef}
+
+import scala.{Array, Boolean, Int, List, Option, None, Nothing, Some, Unit}
+import scala.collection.JavaConverters._
+import scala.util.{Left, Right}
+
+import java.lang.{RuntimeException, SuppressWarnings}
+import java.util.concurrent.ConcurrentHashMap
+
+/**
+ * A coordination mechanism for parallel job management. This
+ * structure provides mechanisms for aggregated notifications
+ * from jobs, deterministic cancelation, "fire and forget"
+ * submission, and sidecar-style cancelation of direct streams.
+ */
+final class JobManager[F[_]: Concurrent: Timer, I, N] private (
+    notificationsQ: Queue[F, Option[N]],
+    dispatchQ: Queue[F, Stream[F, Nothing]]) {
+
+  import JobManager._
+
+  private[this] val meta: ConcurrentHashMap[I, Context[F]] = new ConcurrentHashMap[I, Context[F]]
+
+  val notifications: Stream[F, N] = notificationsQ.dequeue.unNoneTerminate
+
+  /**
+   * Submits a job for parallel execution at the earliest possible moment.
+   * If the job submission queue is full, this function will asynchronously
+   * block until space is available. Note that a job is visible as Pending
+   * immediately upon submission, even when space is unavailable.
+   *
+   * Attempting to submit a job with the same id as a pre-existing job will
+   * produce false.
+   */
+  @SuppressWarnings(Array("org.wartremover.warts.DefaultArguments", "org.wartremover.warts.Equals"))
+  def submit[R](job: Job[F, I, N, R]): F[Boolean] = {
+    val run = job.run
+      .map(_.swap.toOption)
+      .unNone
+      .map(Some(_))
+      .evalMap(notificationsQ.enqueue1)
+      .drain
+
+    val putStatusF = Concurrent[F] delay {
+      val attempt = Context[F](Status.Pending, None)
+      attempt eq meta.putIfAbsent(job.id, attempt)
+    }
+
+    putStatusF flatMap { s =>
+      if (s)
+        dispatchQ.enqueue1(managementMachinery(job.id, run, false)).as(true)
+      else
+        Concurrent[F].pure(false)
+    }
+  }
+
+  /**
+   * Like submit, but produces a managed stream equal to the job's
+   * run stream. Notifications are stripped and routed to the
+   * shared notifications source. The resulting stream is subject
+   * to remote cancelation, the same as any submitted job.
+   *
+   * Attempting to tap a job with a pre-existing id will produce
+   * an error.
+   */
+  def tap[R](job: Job[F, I, N, R]): Stream[F, R] = {
```

Comment:
It's intentional. I guess `jobLimit` is perhaps a bit deceptive. Basically the concept is that if you're using `tap`, you probably have some external limiting mechanism in place, since you're responsible for running the stream yourself.